cpp/src/parquet/level_conversion.cc

```diff
@@ -0,0 +1,170 @@
+
+    valid_bits_writer.Next();
+  }
+  valid_bits_writer.Finish();
+  *values_read = valid_bits_writer.position();
+}
+#endif
+
+template <bool has_repeated_parent>
+void DefinitionLevelsToBitmapSimd(const int16_t* def_levels, int64_t num_def_levels,
+                                  const int16_t required_definition_level,
+                                  int64_t* values_read, int64_t* null_count,
+                                  uint8_t* valid_bits, int64_t valid_bits_offset) {
+  constexpr int64_t kBitMaskSize = 64;
+  ::arrow::internal::FirstTimeBitmapWriter writer(valid_bits,
+                                                  /*start_offset=*/valid_bits_offset,
+                                                  /*length=*/num_def_levels);
+  int64_t set_count = 0;
+  *values_read = 0;
+  while (num_def_levels > 0) {
+    int64_t batch_size = std::min(num_def_levels, kBitMaskSize);
+    CheckLevelRange(def_levels, batch_size, required_definition_level);
+    uint64_t defined_bitmap = internal::GreaterThanBitmap(def_levels, batch_size,
+                                                          required_definition_level - 1);
+    if (has_repeated_parent) {
+#if defined(ARROW_HAVE_BMI2)
+      // This is currently a specialized code path assuming only (nested) lists
+      // present through the leaf (i.e. no structs).
+      // Upper level code only calls this method
+      // when the leaf-values are nullable (otherwise no spacing is needed),
+      // Because only nested lists exists it is sufficient to know that the field
+      // was either null or included it (i.e. definition level > max_definitation_level
+      // -2) If there where structs mixed in, we need to know the def_level of the
+      // repeated parent so we can check for def_level > "def level of repeated parent".
+      uint64_t present_bitmap = internal::GreaterThanBitmap(
+          def_levels, batch_size, required_definition_level - 2);
+      uint64_t selected_bits = _pext_u64(defined_bitmap, present_bitmap);
+      writer.AppendWord(selected_bits, ::arrow::BitUtil::PopCount(present_bitmap));
+      set_count += ::arrow::BitUtil::PopCount(selected_bits);
+#else
+      assert(false && "must not execute this without BMI2");
+#endif
+    } else {
+      writer.AppendWord(defined_bitmap, batch_size);
+      set_count += ::arrow::BitUtil::PopCount(defined_bitmap);
+    }
+    def_levels += batch_size;
+    num_def_levels -= batch_size;
+  }
+
+  *values_read = writer.position();
+  *null_count += *values_read - set_count;
+  writer.Finish();
+}
+
+}  // namespace
+
+void DefinitionLevelsToBitmap(const int16_t* def_levels, int64_t num_def_levels,
+                              const int16_t max_definition_level,
+                              const int16_t max_repetition_level, int64_t* values_read,
+                              int64_t* null_count, uint8_t* valid_bits,
+                              int64_t valid_bits_offset) {
+  if (max_repetition_level > 0) {
```

Comment:
The interlacing of 1 runtime branch and 2 static branches is hard to read. It seems the big-endian version doesn't even close the block with a corresponding parenthesis.

Reply:
split this into two functions.